File: JassToTsPython/Statement.py
import enum

class StatementType(enum.Enum):
    Prog = "Prog"

    TypeDecl   = "TypeDecl"
    Glob       = "Glob"
    GConst     = "GConst"
    GVar       = "GVar"
    GArr       = "GArr"
    LVar       = "LVar"
    LArr       = "LArr"
    Expr       = "Expr"
    RVar       = "RVar"
    FCall      = "FCall"
    RArr       = "RArr"
    RFunc      = "RFunc"
    Native     = "Native"
    CNative    = "CNative"
    FuncDecl   = "FuncDecl"
    Params     = "Params"
    Func       = "Func"
    CFunc      = "CFunc"
    FuncLocals = "FuncLocals"
    FuncBody   = "FuncBody"
    Comm       = "Comm"
    Debug      = "Debug"
    Set        = "Set"
    ASet       = "ASet"
    If         = "If"
    Cond       = "Cond"
    Then       = "Then"
    ElseCond   = "ElseCond"
    Else       = "Else"
    Loop       = "Loop"
    Exit       = "Exit"
    Return     = "Return"

    TypeName = "TypeName"
    BaseType = "BaseType"
    Type     = "Type"
    Name     = "Name"
    Val      = "Val"
    Oper     = "Oper"
    Result   = "Result"
    Param    = "Param"
    Ind      = "Ind"
    Par      = "Par"

    YdweMacro = "YdweMacro"

class Statement:
    def __init__(self, parent = None, type = None, start = None):
        self.Parent = parent
        self.Type = type
        self.Start = start
        self.Childs = []
    
    # Создать и добавить дочернюю инструкцию
    # тип инструкции
    # токен с которого начинается потомок
    # дочернюю инструкцию
    def MakeChild(self, type, token):
        child = Statement(self, type, token)
        self.Childs.append(child)
        return child

    # <summary> Создать и добавить дочернюю инструкцию </summary>
    # <param name="type"> тип инструкции </param>
    # <param name="token"> токен с которого начинается потомок </param>
    # <returns> себя </returns>
    # <seealso cref="MakeChild(string,Token)"/>
    def AddChild(self, type, token):
        self.MakeChild(type, token)
        return self

    # <summary> Добавить дочернюю инструкцию </summary>
    # <param name="child"> дочерняя инструкция </param>
    # <returns> себя </returns>
    def AddChildStatement(self, child):
        child.Parent = self
        self.Childs.append(child)
        return self

    def __str__(self): return self.ToString(0)
    
    def ToString(self, ident) -> str:
        res = "// {}{} {}\n".format("".ljust(ident, ' '), self.Type, self.Start)
        if len(self.Childs) > 0:
            res += "".join(map(lambda x: x.ToString(ident + 2), self.Childs))
                  #"".join(x.ToString(ident + 2) for x in self.Childs)
        return res
            

File: jass-to-ts-go/jass/Statement.go
package jass

import "fmt"

type Statement struct {
	Type   string
	Start  *Token
	Parent *Statement
	Childs []*Statement
}

func NewStatement() *Statement {
	return &Statement{}
}

/**
* Создать и добавить дочернюю инструкцию
*
* @param type  тип инструкции
* @param token токен с которого начинается потомок
* @return дочернюю инструкцию
* @see Statement#AddChild
 */
func (s *Statement) MakeChild(_type string, token *Token) *Statement {
	var child = &Statement{
		_type,
		token,
		s,
		[]*Statement{},
	}
	s.Childs = append(s.Childs, child)
	return child
}

/**
* Создать и добавить дочернюю инструкцию
*
* @param type  тип инструкции
* @param token токен с которого начинается потомок
* @return себя
* @see Statement#MakeChild
 */
func (statement *Statement) AddChild(_type string, token *Token) *Statement {
	statement.MakeChild(_type, token)
	return statement
}

/**
* Добавить дочернюю инструкцию
*
* @param child дочерняя инструкция
* @return себя
 */
func (statement *Statement) AddChildStatement(child *Statement) *Statement {
	child.Parent = statement
	statement.Childs = append(statement.Childs, child)
	return statement
}

func (statement *Statement) String() string {
	return statement.toString(0)
}

func (statement *Statement) toString(ident int) string {
	var s = ""
	for _, statement := range statement.Childs {
		s += statement.toString(ident + 2)
	}
	var sident = fmt.Sprintf("%d", ident)
	sident = fmt.Sprintf("%"+sident+"s", "")
	return fmt.Sprintf("%s%s %s\n%s", sident, statement.Type, statement.Start, s)
}
